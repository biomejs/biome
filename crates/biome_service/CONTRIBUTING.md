# Biome Service

The Biome Service is where we implement the [Workspace](src/workspace.rs).

The workspace is where Biome keeps all internal state of projects, such as open
documents, but also more advanced service data, such as the instances of our
[project layout](../biome_project_layout/) and our
[module graph](../biome_module_graph/).

Note that there are two implementations of the `Workspace` trait:

* [`WorkspaceServer`](src/workspace/server.rs) maintains the state itself, and
  is used both inside the daemon as well as in the CLI when it runs in
  daemonless mode.
* [`WorkspaceClient`](src/workspace/client.rs) is used for creating connections
  to the daemon and communicating with the `WorkspaceServer`.

## Watcher

The state inside our workspace is kept in sync with the file system using the
[`WorkspaceWatcher`](src/workspace_watcher.rs). The watcher is only active in
daemon mode and not used by the CLI.

### Debugging

Debugging the watcher can be tricky, because you need to run the daemon,
interact with the file system, and observe the daemon's state somehow.

Debugging is possible with the VS Code extension or Zed extension,  but for an easier experience,
you can use these CLI commands:

1. Start the daemon using `cargo run --bin=biome -- start`. Note there won't be
   much output to inspect, but the daemon will write its logs in your
   [cache folder](../../crates/biome_fs/src/dir.rs).
2. Run commands against the daemon, such as
   `cargo run --bin=biome -- lint --use-server <path>`.

The rule `noImportCycles` is currently the best candidate to observe the state
in the project layout and the module graph.

### Tests

The watcher has tests related to its workspace methods in
[`watcher.tests.rs`](src/workspace/watcher.tests.rs), but there are also more
end-to-end tests inside the [LSP tests](../biome_lsp/src/server.tests.rs).
