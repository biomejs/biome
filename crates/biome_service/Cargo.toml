[package]
authors.workspace    = true
edition.workspace    = true
license.workspace    = true
name                 = "biome_service"
repository.workspace = true
version              = "0.0.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
biome_analyze        = { workspace = true, features = ["serde"] }
biome_console        = { workspace = true }
biome_deserialize    = { workspace = true }
biome_diagnostics    = { workspace = true }
biome_flags          = { workspace = true }
biome_formatter      = { workspace = true, features = ["serde"] }
biome_fs             = { workspace = true, features = ["serde"] }
biome_js_analyze     = { workspace = true }
biome_js_factory     = { workspace = true, optional = true }
biome_js_formatter   = { workspace = true, features = ["serde"] }
biome_js_parser      = { workspace = true }
biome_js_semantic    = { workspace = true }
biome_js_syntax      = { workspace = true, features = ["serde"] }
biome_json_analyze   = { workspace = true }
biome_json_formatter = { workspace = true }
biome_json_parser    = { workspace = true }
biome_json_syntax    = { workspace = true }
biome_parser         = { workspace = true }
biome_rowan          = { workspace = true, features = ["serde"] }
biome_text_edit      = { workspace = true }
bpaf                 = { workspace = true }
dashmap              = { workspace = true }
indexmap             = { workspace = true, features = ["serde"] }
schemars             = { workspace = true, features = ["indexmap1"], optional = true }
serde                = { workspace = true, features = ["derive"] }
serde_json           = { workspace = true, features = ["raw_value"] }
tracing              = { workspace = true, features = ["attributes"] }

[features]
schema = [
  "dep:schemars",
  "biome_js_analyze/schema",
  "biome_formatter/serde",
  "biome_js_factory",
  "biome_text_edit/schemars",
]

[dev-dependencies]
insta        = { workspace = true }
tests_macros = { workspace = true }
