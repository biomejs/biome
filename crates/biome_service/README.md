# `biome_service`

The crate contains the core functionality of Biome.

## Local installation

```toml
biome_service = { version = "0.0.0", path = "../biome_service" }
```
