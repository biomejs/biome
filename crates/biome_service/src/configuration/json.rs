use crate::configuration::merge::MergeWith;
use crate::configuration::{deserialize_line_width, serialize_line_width, PlainIndentStyle};
use biome_formatter::LineWidth;
use bpaf::Bpaf;
use serde::{Deserialize, Serialize};

/// Options applied to JSON files
#[derive(Default, Debug, Deserialize, Serialize, Eq, PartialEq, Clone, Bpaf)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
#[serde(default, deny_unknown_fields)]
pub struct JsonConfiguration {
    /// Parsing options
    #[serde(skip_serializing_if = "Option::is_none")]
    #[bpaf(external(json_parser), optional)]
    pub parser: Option<JsonParser>,

    /// Formatting options
    #[serde(skip_serializing_if = "Option::is_none")]
    #[bpaf(external(json_formatter), optional)]
    pub formatter: Option<JsonFormatter>,
}

impl JsonConfiguration {
    pub const KNOWN_KEYS: &'static [&'static str] = &["parser", "formatter"];
}

impl MergeWith<JsonConfiguration> for JsonConfiguration {
    fn merge_with(&mut self, other: JsonConfiguration) {
        if let Some(other_parser) = other.parser {
            let parser = self.parser.get_or_insert_with(JsonParser::default);
            parser.merge_with(other_parser);
        }
        if let Some(other_formatter) = other.formatter {
            let formatter = self.formatter.get_or_insert_with(JsonFormatter::default);
            formatter.merge_with(other_formatter);
        }
    }
}

/// Options that changes how the JSON parser behaves
#[derive(Default, Debug, Deserialize, Serialize, Eq, PartialEq, Clone, Bpaf)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
#[serde(rename_all = "camelCase", default, deny_unknown_fields)]
pub struct JsonParser {
    #[bpaf(hide)]
    #[serde(skip_serializing_if = "Option::is_none")]
    /// Allow parsing comments in `.json` files
    pub allow_comments: Option<bool>,
}

impl JsonParser {
    pub(crate) const KNOWN_KEYS: &'static [&'static str] = &["allowComments"];
}

impl MergeWith<JsonParser> for JsonParser {
    fn merge_with(&mut self, other: JsonParser) {
        if let Some(allow_comments) = other.allow_comments {
            self.allow_comments = Some(allow_comments);
        }
    }
}

#[derive(Default, Debug, Deserialize, Serialize, Eq, PartialEq, Clone, Bpaf)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
#[serde(rename_all = "camelCase", default, deny_unknown_fields)]
pub struct JsonFormatter {
    /// Control the formatter for JSON (and its super languages) files.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[bpaf(long("json-formatter-enabled"), argument("true|false"), optional)]
    pub enabled: Option<bool>,

    /// The indent style applied to JSON (and its super languages) files.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[bpaf(long("json-formatter-indent-style"), argument("tab|space"), optional)]
    pub indent_style: Option<PlainIndentStyle>,

    /// The size of the indentation applied to JSON (and its super languages) files. Default to 2.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[bpaf(long("json-formatter-indent-size"), argument("NUMBER"), optional)]
    pub indent_size: Option<u8>,

    /// What's the max width of a line, applied to JSON (and its super languages) files. Defaults to 80.
    #[serde(
        deserialize_with = "deserialize_line_width",
        serialize_with = "serialize_line_width"
    )]
    #[serde(skip_serializing_if = "Option::is_none")]
    #[bpaf(long("json-formatter-line-width"), argument("NUMBER"), optional)]
    pub line_width: Option<LineWidth>,
}

impl JsonFormatter {
    pub(crate) const KNOWN_KEYS: &'static [&'static str] =
        &["enabled", "indentStyle", "indentSize", "lineWidth"];
}

impl MergeWith<JsonFormatter> for JsonFormatter {
    fn merge_with(&mut self, other: JsonFormatter) {
        if let Some(enabled) = other.enabled {
            self.enabled = Some(enabled);
        }
        if let Some(indent_size) = other.indent_size {
            self.indent_size = Some(indent_size);
        }
        if let Some(indent_style) = other.indent_style {
            self.indent_style = Some(indent_style);
        }
        if let Some(line_width) = other.line_width {
            self.line_width = Some(line_width);
        }
    }
}

impl MergeWith<Option<JsonFormatter>> for JsonConfiguration {
    fn merge_with(&mut self, other: Option<JsonFormatter>) {
        if let Some(other_formatter) = other {
            let formatter = self.formatter.get_or_insert_with(JsonFormatter::default);
            formatter.merge_with(other_formatter);
        }
    }
}
