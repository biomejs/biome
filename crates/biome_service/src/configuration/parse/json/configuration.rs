use crate::Configuration;
use biome_deserialize::{
    Deserializable, DeserializableValue, DeserializationDiagnostic, DeserializationVisitor,
    NoneState, Text, VisitableType,
};

impl Deserializable for Configuration {
    fn deserialize(
        value: &impl DeserializableValue,
        name: &str,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self> {
        value.deserialize(ConfigurationVisitor, name, diagnostics)
    }
}

struct ConfigurationVisitor;
impl DeserializationVisitor for ConfigurationVisitor {
    type Output = Configuration;

    const EXPECTED_TYPE: VisitableType = VisitableType::MAP;

    fn visit_map(
        self,
        members: impl Iterator<Item = Option<(impl DeserializableValue, impl DeserializableValue)>>,
        _range: biome_rowan::TextRange,
        _name: &str,
        diagnostics: &mut Vec<DeserializationDiagnostic>,
    ) -> Option<Self::Output> {
        const ALLOWED_KEYS: &[&str] = &[
            "vcs",
            "files",
            "linter",
            "formatter",
            "javascript",
            "json",
            "css",
            "$schema",
            "organizeImports",
            "extends",
            "overrides",
        ];
        let mut result = Self::Output::none();
        for (key, value) in members.flatten() {
            let Some(key_text) = Text::deserialize(&key, "", diagnostics) else {
                continue;
            };
            match key_text.text() {
                "$schema" => {
                    result.schema = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "files" => {
                    result.files = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "vcs" => {
                    result.vcs = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "formatter" => {
                    result.formatter = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "linter" => {
                    result.linter = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "javascript" => {
                    result.javascript = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "json" => {
                    result.json = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "css" => {
                    result.css = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "organizeImports" => {
                    result.organize_imports =
                        Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "extends" => {
                    result.extends = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                "overrides" => {
                    result.overrides = Deserializable::deserialize(&value, &key_text, diagnostics);
                }
                unknown_key => diagnostics.push(DeserializationDiagnostic::new_unknown_key(
                    unknown_key,
                    key.range(),
                    ALLOWED_KEYS,
                )),
            }
        }
        Some(result)
    }
}
