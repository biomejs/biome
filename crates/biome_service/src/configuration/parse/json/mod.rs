//! This module is responsible to parse the configuration from a JSON format
//!

mod configuration;
mod files;
mod formatter;
mod javascript;
mod json_impl;
mod linter;
mod organize_imports;
mod overrides;
mod rules;
mod vcs;
