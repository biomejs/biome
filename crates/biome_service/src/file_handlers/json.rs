use super::{CodeActionsParams, DocumentFileSource, ExtensionHandler, Mime, ParseResult};
use crate::configuration::{to_analyzer_rules, PartialConfiguration};
use crate::file_handlers::DebugCapabilities;
use crate::file_handlers::{
    AnalyzerCapabilities, Capabilities, FixAllParams, FormatterCapabilities, LintParams,
    LintResults, ParserCapabilities,
};
use crate::settings::{
    FormatSettings, Language, LanguageListSettings, LanguageSettings, OverrideSettings,
    SettingsHandle,
};
use crate::workspace::{
    FixFileResult, GetSyntaxTreeResult, OrganizeImportsResult, PullActionsResult,
};
use crate::WorkspaceError;
use biome_analyze::options::PreferredQuote;
use biome_analyze::{
    AnalysisFilter, AnalyzerConfiguration, AnalyzerOptions, ControlFlow, Never, RuleCategories,
};
use biome_deserialize::json::deserialize_from_json_ast;
use biome_diagnostics::{category, Diagnostic, DiagnosticExt, Severity};
use biome_formatter::{FormatError, IndentStyle, IndentWidth, LineEnding, LineWidth, Printed};
use biome_fs::{BiomePath, ConfigName, ROME_JSON};
use biome_json_analyze::analyze;
use biome_json_formatter::context::{JsonFormatOptions, TrailingCommas};
use biome_json_formatter::format_node;
use biome_json_parser::JsonParserOptions;
use biome_json_syntax::{JsonLanguage, JsonRoot, JsonSyntaxNode};
use biome_parser::AnyParse;
use biome_rowan::{AstNode, NodeCache};
use biome_rowan::{TextRange, TextSize, TokenAtOffset};
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Default, serde::Serialize, serde::Deserialize)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
pub struct JsonFormatterSettings {
    pub line_ending: Option<LineEnding>,
    pub line_width: Option<LineWidth>,
    pub indent_width: Option<IndentWidth>,
    pub indent_style: Option<IndentStyle>,
    pub trailing_commas: Option<TrailingCommas>,
    pub enabled: Option<bool>,
}

#[derive(Debug, Clone, Default, serde::Serialize, serde::Deserialize)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
pub struct JsonParserSettings {
    pub allow_comments: bool,
    pub allow_trailing_commas: bool,
}

impl Language for JsonLanguage {
    type FormatterSettings = JsonFormatterSettings;
    type LinterSettings = ();
    type OrganizeImportsSettings = ();
    type FormatOptions = JsonFormatOptions;
    type ParserSettings = JsonParserSettings;
    fn lookup_settings(language: &LanguageListSettings) -> &LanguageSettings<Self> {
        &language.json
    }

    fn resolve_format_options(
        global: &FormatSettings,
        overrides: &OverrideSettings,
        language: &JsonFormatterSettings,
        path: &BiomePath,
        _document_file_source: &DocumentFileSource,
    ) -> Self::FormatOptions {
        let indent_style = if let Some(indent_style) = language.indent_style {
            indent_style
        } else {
            global.indent_style.unwrap_or_default()
        };
        let line_width = if let Some(line_width) = language.line_width {
            line_width
        } else {
            global.line_width.unwrap_or_default()
        };
        let indent_width = if let Some(indent_width) = language.indent_width {
            indent_width
        } else {
            global.indent_width.unwrap_or_default()
        };

        let line_ending = if let Some(line_ending) = language.line_ending {
            line_ending
        } else {
            global.line_ending.unwrap_or_default()
        };

        overrides.override_json_format_options(
            path,
            JsonFormatOptions::new()
                .with_line_ending(line_ending)
                .with_indent_style(indent_style)
                .with_indent_width(indent_width)
                .with_line_width(line_width)
                .with_trailing_commas(language.trailing_commas.unwrap_or_default()),
        )
    }
}

#[derive(Debug, Default, PartialEq, Eq)]
pub(crate) struct JsonFileHandler;

impl ExtensionHandler for JsonFileHandler {
    fn mime(&self) -> super::Mime {
        Mime::Json
    }

    fn may_use_tabs(&self) -> bool {
        true
    }

    fn capabilities(&self) -> Capabilities {
        Capabilities {
            parser: ParserCapabilities { parse: Some(parse) },
            debug: DebugCapabilities {
                debug_syntax_tree: Some(debug_syntax_tree),
                debug_control_flow: None,
                debug_formatter_ir: Some(debug_formatter_ir),
            },
            analyzer: AnalyzerCapabilities {
                lint: Some(lint),
                code_actions: Some(code_actions),
                rename: None,
                fix_all: Some(fix_all),
                organize_imports: Some(organize_imports),
            },
            formatter: FormatterCapabilities {
                format: Some(format),
                format_range: Some(format_range),
                format_on_type: Some(format_on_type),
            },
        }
    }
}

fn is_file_allowed(path: &Path) -> bool {
    path.file_name()
        .and_then(|f| f.to_str())
        .map(|f| {
            super::DocumentFileSource::KNOWN_FILES_AS_JSONC
                .binary_search(&f)
                .is_ok()
        })
        // default is false
        .unwrap_or_default()
}

fn parse(
    biome_path: &BiomePath,
    mut file_source: DocumentFileSource,
    text: &str,
    settings: SettingsHandle,
    cache: &mut NodeCache,
) -> ParseResult {
    let parser = &settings.as_ref().languages.json.parser;
    let overrides = &settings.as_ref().override_settings;
    let options: JsonParserOptions = overrides.override_json_parser_options(
        biome_path,
        JsonParserOptions {
            allow_comments: parser.allow_comments
                || file_source.to_json_file_source().map(|fs| fs.is_jsonc()) == Some(true)
                || is_file_allowed(biome_path),
            allow_trailing_commas: parser.allow_trailing_commas || is_file_allowed(biome_path),
        },
    );
    if let DocumentFileSource::Json(file_source) = &mut file_source {
        file_source.set_allow_trailing_comma(options.allow_trailing_commas);
    }
    let parse = biome_json_parser::parse_json_with_cache(text, cache, options);
    let root = parse.syntax();
    let diagnostics = parse.into_diagnostics();

    ParseResult {
        any_parse: AnyParse::new(
            // SAFETY: the parser should always return a root node
            root.as_send().unwrap(),
            diagnostics,
        ),
        language: Some(file_source),
    }
}

fn debug_syntax_tree(_rome_path: &BiomePath, parse: AnyParse) -> GetSyntaxTreeResult {
    let syntax: JsonSyntaxNode = parse.syntax();
    let tree: JsonRoot = parse.tree();
    GetSyntaxTreeResult {
        cst: format!("{syntax:#?}"),
        ast: format!("{tree:#?}"),
    }
}

fn debug_formatter_ir(
    path: &BiomePath,
    document_file_source: &DocumentFileSource,
    parse: AnyParse,
    settings: SettingsHandle,
) -> Result<String, WorkspaceError> {
    let options = settings.format_options::<JsonLanguage>(path, document_file_source);

    let tree = parse.syntax();
    let formatted = format_node(options, &tree)?;

    let root_element = formatted.into_document();
    Ok(root_element.to_string())
}

#[tracing::instrument(level = "debug", skip(parse, settings))]
fn format(
    path: &BiomePath,
    document_file_source: &DocumentFileSource,
    parse: AnyParse,
    settings: SettingsHandle,
) -> Result<Printed, WorkspaceError> {
    let options = settings.format_options::<JsonLanguage>(path, document_file_source);

    tracing::debug!("Format with the following options: \n{}", options);

    let tree = parse.syntax();
    let formatted = format_node(options, &tree)?;

    match formatted.print() {
        Ok(printed) => Ok(printed),
        Err(error) => Err(WorkspaceError::FormatError(error.into())),
    }
}

fn format_range(
    path: &BiomePath,
    document_file_source: &DocumentFileSource,
    parse: AnyParse,
    settings: SettingsHandle,
    range: TextRange,
) -> Result<Printed, WorkspaceError> {
    let options = settings.format_options::<JsonLanguage>(path, document_file_source);

    let tree = parse.syntax();
    let printed = biome_json_formatter::format_range(options, &tree, range)?;
    Ok(printed)
}

fn format_on_type(
    path: &BiomePath,
    document_file_source: &DocumentFileSource,
    parse: AnyParse,
    settings: SettingsHandle,
    offset: TextSize,
) -> Result<Printed, WorkspaceError> {
    let options = settings.format_options::<JsonLanguage>(path, document_file_source);

    let tree = parse.syntax();

    let range = tree.text_range();
    if offset < range.start() || offset > range.end() {
        return Err(WorkspaceError::FormatError(FormatError::RangeError {
            input: TextRange::at(offset, TextSize::from(0)),
            tree: range,
        }));
    }

    let token = match tree.token_at_offset(offset) {
        // File is empty, do nothing
        TokenAtOffset::None => panic!("empty file"),
        TokenAtOffset::Single(token) => token,
        // The cursor should be right after the closing character that was just typed,
        // select the previous token as the correct one
        TokenAtOffset::Between(token, _) => token,
    };

    let root_node = match token.parent() {
        Some(node) => node,
        None => panic!("found a token with no parent"),
    };

    let printed = biome_json_formatter::format_sub_tree(options, &root_node)?;
    Ok(printed)
}

fn lint(params: LintParams) -> LintResults {
    tracing::debug_span!("Linting JSON file", path =? params.path, language =? params.language)
        .in_scope(move || {
            let root: JsonRoot = params.parse.tree();
            let mut diagnostics = params.parse.into_diagnostics();
            let settings = params.settings.as_ref();

            // if we're parsing the `biome.json` file, we deserialize it, so we can emit diagnostics for
            // malformed configuration
            if params.path.ends_with(ROME_JSON)
                || params.path.ends_with(ConfigName::biome_json())
                || params.path.ends_with(ConfigName::biome_jsonc())
            {
                let deserialized = deserialize_from_json_ast::<PartialConfiguration>(&root, "");
                diagnostics.extend(
                    deserialized
                        .into_diagnostics()
                        .into_iter()
                        .map(biome_diagnostics::serde::Diagnostic::new)
                        .collect::<Vec<_>>(),
                );
            }

            let mut diagnostic_count = diagnostics.len() as u32;
            let mut errors = diagnostics
                .iter()
                .filter(|diag| diag.severity() <= Severity::Error)
                .count();

            let skipped_diagnostics = diagnostic_count - diagnostics.len() as u32;

            let rules = settings.as_rules(params.path.as_path());
            let rule_filter_list = rules
                .as_ref()
                .map(|rules| rules.as_enabled_rules())
                .unwrap_or_default()
                .into_iter()
                .collect::<Vec<_>>();

            let analyzer_options =
                compute_analyzer_options(&params.settings, PathBuf::from(params.path.as_path()));
            let mut filter = AnalysisFilter::from_enabled_rules(Some(rule_filter_list.as_slice()));
            filter.categories = params.categories;
            let has_lint = filter.categories.contains(RuleCategories::LINT);

            let (_, analyze_diagnostics) = analyze(&root, filter, &analyzer_options, |signal| {
                if let Some(mut diagnostic) = signal.diagnostic() {
                    // Do not report unused suppression comment diagnostics if this is a syntax-only analyzer pass
                    if !has_lint && diagnostic.category() == Some(category!("suppressions/unused"))
                    {
                        return ControlFlow::<Never>::Continue(());
                    }

                    diagnostic_count += 1;

                    // We do now check if the severity of the diagnostics should be changed.
                    // The configuration allows to change the severity of the diagnostics emitted by rules.
                    let severity = diagnostic
                        .category()
                        .filter(|category| category.name().starts_with("lint/"))
                        .map_or_else(
                            || diagnostic.severity(),
                            |category| {
                                rules
                                    .as_ref()
                                    .and_then(|rules| rules.get_severity_from_code(category))
                                    .unwrap_or(Severity::Warning)
                            },
                        );

                    if severity <= Severity::Error {
                        errors += 1;
                    }

                    if diagnostic_count <= params.max_diagnostics {
                        for action in signal.actions() {
                            if !action.is_suppression() {
                                diagnostic = diagnostic.add_code_suggestion(action.into());
                            }
                        }

                        let error = diagnostic.with_severity(severity);

                        diagnostics.push(biome_diagnostics::serde::Diagnostic::new(error));
                    }
                }

                ControlFlow::<Never>::Continue(())
            });

            diagnostics.extend(
                analyze_diagnostics
                    .into_iter()
                    .map(biome_diagnostics::serde::Diagnostic::new)
                    .collect::<Vec<_>>(),
            );

            LintResults {
                diagnostics,
                errors,
                skipped_diagnostics,
            }
        })
}

fn code_actions(_: CodeActionsParams) -> PullActionsResult {
    PullActionsResult {
        actions: Vec::new(),
    }
}

fn fix_all(params: FixAllParams) -> Result<FixFileResult, WorkspaceError> {
    let tree: JsonRoot = params.parse.tree();
    Ok(FixFileResult {
        actions: vec![],
        errors: 0,
        skipped_suggested_fixes: 0,
        code: tree.syntax().to_string(),
    })
}

fn organize_imports(parse: AnyParse) -> Result<OrganizeImportsResult, WorkspaceError> {
    Ok(OrganizeImportsResult {
        code: parse.syntax::<JsonLanguage>().to_string(),
    })
}

fn compute_analyzer_options(settings: &SettingsHandle, file_path: PathBuf) -> AnalyzerOptions {
    let configuration = AnalyzerConfiguration {
        rules: to_analyzer_rules(settings.as_ref(), file_path.as_path()),
        globals: vec![],
        preferred_quote: PreferredQuote::Double,
    };
    AnalyzerOptions {
        configuration,
        file_path,
    }
}
