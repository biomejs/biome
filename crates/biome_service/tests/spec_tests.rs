use biome_deserialize::json::deserialize_from_json_str;
use biome_diagnostics::{print_diagnostic_to_string, DiagnosticExt};
use biome_json_parser::JsonParserOptions;
use biome_service::Configuration;
use std::ffi::OsStr;
use std::fs::read_to_string;
use std::path::Path;

tests_macros::gen_tests! {"tests/invalid/*.{json}", crate::run_invalid_configurations, "module"}
tests_macros::gen_tests! {"tests/valid/*.{json}", crate::run_valid_configurations, "module"}

fn run_invalid_configurations(input: &'static str, _: &str, _: &str, _: &str) {
    let input_file = Path::new(input);
    let file_name = input_file.file_name().and_then(OsStr::to_str).unwrap();
    let extension = input_file.extension().and_then(OsStr::to_str).unwrap();
    let input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {:?}: {:?}", input_file, err));

    let result = match extension {
        "json" => deserialize_from_json_str::<Configuration>(
            input_code.as_str(),
            JsonParserOptions::default(),
        ),
        "jsonc" => deserialize_from_json_str::<Configuration>(
            input_code.as_str(),
            JsonParserOptions::default().with_allow_comments(),
        ),
        _ => {
            panic!("Extension not supported");
        }
    };

    assert!(
        result.has_errors(),
        "This test should have diagnostics, but it doesn't have any"
    );

    let result = result
        .into_diagnostics()
        .into_iter()
        .map(|diagnostic| {
            print_diagnostic_to_string(
                &diagnostic
                    .with_file_path(file_name)
                    .with_file_source_code(input_code.as_str()),
            )
        })
        .collect::<Vec<_>>()
        .join("\n\n");

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(file_name, result, file_name);
    });
}

fn run_valid_configurations(input: &'static str, _: &str, _: &str, _: &str) {
    let input_file = Path::new(input);
    let file_name = input_file.file_name().and_then(OsStr::to_str).unwrap();
    let extension = input_file.extension().and_then(OsStr::to_str).unwrap();
    let input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {:?}: {:?}", input_file, err));

    let result = match extension {
        "json" => deserialize_from_json_str::<Configuration>(
            input_code.as_str(),
            JsonParserOptions::default(),
        ),
        "jsonc" => deserialize_from_json_str::<Configuration>(
            input_code.as_str(),
            JsonParserOptions::default().with_allow_comments(),
        ),
        _ => {
            panic!("Extension not supported");
        }
    };

    let has_errors = result.has_errors();

    if has_errors {
        let diagnostics = result
            .into_diagnostics()
            .into_iter()
            .map(|diagnostic| {
                print_diagnostic_to_string(
                    &diagnostic
                        .with_file_path(file_name)
                        .with_file_source_code(input_code.as_str()),
                )
            })
            .collect::<Vec<_>>()
            .join("\n\n");
        panic!(
            "This test should not have diagnostics, but some have been emitted.\n {}",
            diagnostics
        );
    } else {
        // test passed, no errors
    }
}

#[ignore]
#[test]
fn quick_test() {
    let source = r#"{
        "javascript": {
            "formatter": {
                "overrides": [
                {}]
            }
        }
    }"#;
    let result = deserialize_from_json_str::<Configuration>(source, JsonParserOptions::default());

    dbg!(result.diagnostics());
    assert!(!result.has_errors());
}
