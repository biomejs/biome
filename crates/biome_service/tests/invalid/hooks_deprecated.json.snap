---
source: crates/biome_service/tests/spec_tests.rs
expression: hooks_deprecated.json
---
hooks_deprecated.json:9:16 deserialize  DEPRECATED  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! The property hooks is deprecated.
  
     7 │ 					"level": "error",
     8 │ 					"options": {
   > 9 │ 						"hooks": [
       │ 						         ^
  > 10 │ 							{
        ...
  > 15 │ 							{ "name": "useQuery", "closureIndex": 1, "dependenciesIndex": 0 }
  > 16 │ 						]
       │ 						^
    17 │ 					}
    18 │ 				}
  
  i useHookAtTopLevel now uses the React hook naming convention to determine hook calls.
  


